src/utils/auth.rs

```diff
@@ -238,21 +150,247 @@ fn hmac_sha256(key: &[u8], message: &[u8]) -> Vec<u8> {
     result.into_bytes().to_vec()
 }
 
+fn create_canonical_request(
+    method: &str,
+    path: &str,
+    headers: &HeaderMap,
+    signed_headers: Vec<&str>,
+    query_string: &str,
+    body: &BytesMut,
+    content_hash: &str,
+) -> String {
+    let decoded_path = percent_decode_str(path).decode_utf8().unwrap();
+    
+    // Check for CLI version based on headers
+    let is_v2_23 = headers.contains_key("x-amz-sdk-checksum-algorithm");
+    
+    // Get canonical headers and signed headers based on version
+    let (canonical_headers, final_signed_headers) = if is_v2_23 {
+        get_v2_23_headers(headers, &signed_headers)
+    } else {
+        get_v2_22_headers(headers, &signed_headers)
+    };
+
+    // Handle payload hash based on version and headers
+    let payload_hash = if content_hash == "UNSIGNED-PAYLOAD" {
+        content_hash.to_string()
+    } else if is_v2_23 {
+        // Use the provided content hash for v2.23
+        content_hash.to_string()
```

Comment:
Why are you not hashing the payload if the payload is signed? This introduces a security vulnerability where you are not validating that the payload hash matches the hash provided in the signature.